<launch>
  <arg name="n_trials" default="3" />
  <arg name="bagfiles" default="/data/stewart/airsim-bags/reef-explore_3_part_1_of_5.bag  /data/stewart/airsim-bags/reef-explore_3_part_2_of_5.bag  /data/stewart/airsim-bags/reef-explore_3_part_3_of_5.bag
/data/stewart/airsim-bags/reef-explore_3_part_4_of_5.bag  /data/stewart/airsim-bags/reef-explore_3_part_5_of_5.bag" />

  <arg name="image_topic" default="/camera/rgb/image_color" /> <!--Rectified color image-->
  <arg name="depth_cloud_topic" default="/camera/points" /> <!--Depth cloud topic-->
  <arg name="segmentation_topic" default="/camera/seg/image_color" /> <!--Rectified color image-->
  <arg name="file_prefix" default="" />
  <arg name="output_dir" default="" />
  <arg name="match_methods" default="id,hungarian-l1,clear-l1" />

  <!-- Vision arguments -->
  <arg name="transform_topic" default="" />
  <arg name="world_frame" default="map" />
  <arg name="sensor_frame" default="bottom_camera" />
  <arg name="scale" default="1.0" />
  <arg name="rate" default="3" />
  <arg name="hue" default="true" />
  <arg name="intensity" default="true" />
  <arg name="clahe" default="false" />
  <arg name="texton" default="false" />
  <arg name="orb" default="true" />
  <arg name="surf" default="false" />
  <arg name="use_pc" default="true" />s
  <arg name="use_tf" default="true" />

  <!-- Topic model arguments -->
  <arg name="K" default="10" />
  <arg name="V" default="15436" />
  <arg name="alpha" default="0.3" />
  <arg name="beta" default="0.04" />
  <arg name="gamma" default="0.0" />
  <arg name="tau" default="0.5" />
  <arg name="p_refine_last_observation" default="0.5" />
  <arg name="num_threads" default="2" />
  <arg name="cell_space" default="0.8" />
  <arg name="cell_time" default="3600" />
  <arg name="cell_size" default="" />
  <arg name="G_time" default="1" />
  <arg name="G_space" default="1" />
  <arg name="polled_refine" default="false" />
  <arg name="words_topic" default="/rost_vision/words" />
  <arg name="min_obs_refine_time" default="600" />

  <arg name="ppx_topic" default="/ppx_cloud" />
  <arg name="map_type" default="TopicMap" />
  <arg name="show_topics" default="false" />
  <arg name="2d_words_topic" default="/rost_vision/words_2d" />
  <arg name="show_words" default="true" />
  <arg name="show_perplexity" default="false" />
  <arg name="show_equalized" default="true" />
  <arg name="ppx_display_factor" default="0.75" />

  <arg name="use_rviz" default="false" />
  <arg name="visualize_3d" default="$(arg use_rviz)" />

  <param name="/use_sim_time" value="false"/>


  <node pkg="sunshine" type="utils.multi_agent_match" output="screen" name="multi_agent_sim" args="record $(arg n_trials) $(arg bagfiles)">
    <param name="image_topic" value="$(arg image_topic)" />
    <param name="depth_cloud_topic" value="$(arg depth_cloud_topic)" />
    <param name="segmentation_topic" value="$(arg segmentation_topic)" />

    <param name="K" type="int" value="$(arg K)" />
    <param name="V" type="int" value="$(arg V)" />
    <param name="alpha" type="double" value="$(arg alpha)" />
    <param name="beta" type="double" value="$(arg beta)" />
    <param name="gamma" type="double" value="$(arg gamma)" />
    <param name="tau" type="double" value="$(arg tau)" />
    <param name="p_refine_last_observation" type="double" value="$(arg p_refine_last_observation)" />
    <param name="num_threads" type="int" value="$(arg num_threads)" />
    <param name="cell_size" type="string" value="$(arg cell_size)" />
    <param name="cell_space" type="double" value="$(arg cell_space)" />
    <param name="cell_time" type="double" value="$(arg cell_time)" />
    <param name="G_time" type="int" value="$(arg G_time)" />
    <param name="G_space" type="int" value="$(arg G_space)" />
    <param name="polled_refine" value="$(arg polled_refine)" />
    <param name="words_topic" value="$(arg words_topic)" />
    <param name="min_obs_refine_time" value="$(arg min_obs_refine_time)" />

    <param name="feature_descriptor" type="string" value="ORB"/>
    <param name="scale" type="double" value="$(arg scale)"/>
    <param name="rate" type="double" value="$(arg rate)"/>
    <param name="use_clahe" value="$(arg clahe)" />
    <param name="use_hue" value="$(arg hue)"/>
    <param name="use_intensity" value="$(arg intensity)"/>
    <param name="use_orb" value="$(arg orb)"/>
    <param name="use_surf" value="$(arg surf)"/>
    <param name="use_texton" value="$(arg texton)" />
    <param name="image" value="$(arg image_topic)"/>
    <param name="transform" value="$(arg transform_topic)"/>
    <param name="sensor_frame" value="$(arg sensor_frame)" />
    <param name="world_frame" value="$(arg world_frame)" />
    <param name="pc" value="$(arg depth_cloud_topic)" />
    <param name="use_pc" value="$(arg use_pc)" />
    <param name="use_tf" value="$(arg use_tf)" />
    <param name="publish_2d_words" value="true" />
    <param name="output_prefix" value="$(arg output_dir)" />
    <param name="box" value="-50x-50x100x100" />
    <param name="match_methods" value="$(arg match_methods)" />
  </node>

<!--   <node pkg="sunshine" type="visualize3d" name="visualize_0" if="$(arg visualize_3d)"> -->
<!--     <param name="input_topic" type="string" value="/0/map" /> -->
<!--     <param name="input_type" type="string" value="$(arg map_type)" /> -->
<!--     <param name="output_topic" type="string" value="/0/topic_cloud" /> -->
<!--     <param name="ppx_topic" type="string" value="$(arg ppx_topic)" /> -->
<!--     <param name="ppx_display_factor" type="double" value="$(arg ppx_display_factor)" /> -->
<!--     <param name="world_frame" type="string" value="$(arg world_frame)" /> -->
<!--     <param name="cell_size" type="string" value="$(arg cell_size)" /> -->
<!--     <param name="cell_space" type="double" value="$(arg cell_space)" /> -->
<!--     <param name="cell_time" type="double" value="$(arg cell_time)" /> -->
<!--   </node> -->

<!--   <node pkg="sunshine" type="visualize3d" name="visualize_1" if="$(arg visualize_3d)"> -->
<!--     <param name="input_topic" type="string" value="/1/map" /> -->
<!--     <param name="input_type" type="string" value="$(arg map_type)" /> -->
<!--     <param name="output_topic" type="string" value="/1/topic_cloud" /> -->
<!--     <param name="ppx_topic" type="string" value="$(arg ppx_topic)" /> -->
<!--     <param name="ppx_display_factor" type="double" value="$(arg ppx_display_factor)" /> -->
<!--     <param name="world_frame" type="string" value="$(arg world_frame)" /> -->
<!--     <param name="cell_size" type="string" value="$(arg cell_size)" /> -->
<!--     <param name="cell_space" type="double" value="$(arg cell_space)" /> -->
<!--     <param name="cell_time" type="double" value="$(arg cell_time)" /> -->
<!--   </node> -->

  <node pkg="sunshine" type="visualize3d" name="visualize_naive" if="$(arg visualize_3d)">
    <param name="input_topic" type="string" value="/naive_map" />
    <param name="input_type" type="string" value="$(arg map_type)" />
    <param name="output_topic" type="string" value="/naive_topic_cloud" />
    <param name="ppx_topic" type="string" value="$(arg ppx_topic)" />
    <param name="ppx_display_factor" type="double" value="$(arg ppx_display_factor)" />
    <param name="world_frame" type="string" value="$(arg world_frame)" />
    <param name="cell_size" type="string" value="$(arg cell_size)" />
    <param name="cell_space" type="double" value="$(arg cell_space)" />
    <param name="cell_time" type="double" value="$(arg cell_time)" />
  </node>

  <node pkg="sunshine" type="visualize3d" name="visualize_merged" if="$(arg visualize_3d)">
    <param name="input_topic" type="string" value="/merged_map" />
    <param name="input_type" type="string" value="$(arg map_type)" />
    <param name="output_topic" type="string" value="/merged_topic_cloud" />
    <param name="ppx_topic" type="string" value="$(arg ppx_topic)" />
    <param name="ppx_display_factor" type="double" value="$(arg ppx_display_factor)" />
    <param name="world_frame" type="string" value="$(arg world_frame)" />
    <param name="cell_size" type="string" value="$(arg cell_size)" />
    <param name="cell_space" type="double" value="$(arg cell_space)" />
    <param name="cell_time" type="double" value="$(arg cell_time)" />
  </node>

  <node pkg="sunshine" type="visualize3d" name="visualize_gt" if="$(arg visualize_3d)">
    <param name="input_topic" type="string" value="/gt_map" />
    <param name="input_type" type="string" value="$(arg map_type)" />
    <param name="output_topic" type="string" value="/gt_topic_cloud" />
    <param name="ppx_topic" type="string" value="$(arg ppx_topic)" />
    <param name="ppx_display_factor" type="double" value="$(arg ppx_display_factor)" />
    <param name="world_frame" type="string" value="$(arg world_frame)" />
    <param name="cell_size" type="string" value="$(arg cell_size)" />
    <param name="cell_space" type="double" value="$(arg cell_space)" />
    <param name="cell_time" type="double" value="$(arg cell_time)" />
  </node>

  <node type="rviz" name="rviz" pkg="rviz" args="-d $(find sunshine)/rviz/multi_agent_topics.rviz" if="$(arg use_rviz)" />
</launch>
